### ISSUE TITLE
Continue button has stopped working on tutorials deployed to shinyapps.io

### ISSUE THREAD

**Issue by garrettgman** [rstudio/learnr#50]
Continue buttons do not work for apps that I deploy to shinyapps.io, e.g.

https://tutorials.shinyapps.io/00-setup/
Source code: https://github.com/rstudio/ggtutorials/blob/master/tutorials/00-setup/00-setup.Rmd

This began after I updated tutor from github this morning.
The  continue buttons work when I run the apps locally.

I noticed this while updating tutorials to use as examples in the documentation.

**Comment by jjallaire** [comment]
I think the problem has to do with shiny_prerendered, particularly that we
don't re-render the Rmd file if it hasn't changed. This means that when we
update tutor with new features you need to force a re-rendering of your
tutorials before deploying them. You can read about this here:

http://rmarkdown.rstudio.com/authoring_shiny_prerendered.html#forcing-a-render

The easiest thing to do is probably to just remove all the .html files and
re-render the tutorials. Another approach is to just manually call
rmarkdown::render on each of the tutorials.

**Comment by garrettgman** [comment]
@jjallaire That seems to be the source, but re-rendering and redeploying did not fix the Continue button until I deleted the app from my shinyapps.io account and then redeployed it (a heavy handed attempt to clear the files on shinyapps.io). Is there a better way to make sure that the app is being re-rendered on the shinyapps.io server?

**Comment by garrettgman** [comment]
@jjallaire I'm no longer sure that this is a shiny-prerender issue. Re-rendering and redeploying has failed to activate the Continue button on several more apps (even on apps that I delete from shinyapps.io before redeploying).

Instead, this reliably fixes the problem: after deploying the app and verifying that the Continue buttons do not work, I click "Start Over". Then the buttons begin to work.

**Comment by jjallaire** [comment]
If there is a tutorial where you can still reproduce this would you mind
capturing the JavaScript log both on page load and when you click the
Continue button?

**Comment by jjallaire** [comment]
Or just point me to a URL of a tutorial that exhibits this behavior

**Comment by garrettgman** [comment]
All of the example tutorials seem to display the behavior when you visit them the first time. Here are some screenshots of https://tutorials.shinyapps.io/00-setup/ with the JavaScript console (is this the log you have in mind?)

Immediately after page load

![screen shot 2017-03-17 at 12 58 11 pm](https://cloud.githubusercontent.com/assets/85014/24054299/1e932b40-0b12-11e7-853f-381cfc6bd647.png)

After clicking "Continue" (no change)

![screen shot 2017-03-17 at 12 58 39 pm](https://cloud.githubusercontent.com/assets/85014/24054308/2817538a-0b12-11e7-9068-bfedaa17fcdc.png)

After clicking "Start Over"

![screen shot 2017-03-17 at 12 58 54 pm](https://cloud.githubusercontent.com/assets/85014/24054329/396ba85c-0b12-11e7-9956-13994c233423.png)

After clicking "Continue" a second time

![screen shot 2017-03-17 at 12 58 59 pm](https://cloud.githubusercontent.com/assets/85014/24054347/43df69c2-0b12-11e7-8890-c7ef2e570d3d.png)

Note that some of the tutorials begin with a Welcome section (`##`), but no sub-section (`###`) in the first topic. In these tutorials, the first Continue button is a proxy "Next Topic" button and works as expected. However, the true "Continue" buttons, which advance between sub-section do not work.

**Comment by robbyshaver** [comment]
So if you press "Start Over" everything works fine? If that's the case, I suspect this issue was caused by the recent code change of skipping sections instead of exercises.

**Comment by garrettgman** [comment]
That seems to be correct: pressing "Start Over" fixes the behavior for that
session.

If you visit this tutorial, there may be a second clue:
https://tutorials.shinyapps.io/01-data-basics/
* the first "Continue" button is a proxy for a "next Topic" button and works
* After you click through, the second "Continue" button is a true Continue
button that progressively reveals a sub-section. It doesn't work.
* When you click "Start Over" to resolve the problem, the First "Continue"
button changes into a blue "Next Topic" button.

**Comment by jjallaire** [comment]
It sounds like though that this error occurs for even newly deployed tutorials (i.e. it happens until there is some history to replay at startup, if there is no history at startup then we get the JS error seen above). @robbyshaver Do you have enough to go on here?

**Comment by jjallaire** [comment]
@robbyshaver I can confirm that if you visit a tutorial for the very first time you will get that JavaScript exception (it goes away after "Start Over"). My guess is that your code is implicitly assuming that some history will have been played back from the server and when that isn't the case there is a data structure that isn't populated the way you expect.

**Comment by robbyshaver** [comment]
Here's the error I'm seeing in the console. It's old code - the exerciseSkipped function doesn't exist anymore.

![image](https://cloud.githubusercontent.com/assets/1553877/24063456/19c8dbea-0b36-11e7-8371-c618c6ffd6e1.png)

**Comment by jjallaire** [comment]
Given that it's old code, Garrett there might be a problem with the machine
you are deploying from having an older version of the package.
Could you make sure to install from github and then try again?

**Comment by garrettgman** [comment]
Thanks @robbyshaver. It looks like there have been several pull requests since I updated tutor this morning, one involving `exercise_skipper`. I reinstalled tutor (right now), re-rendered each tutorial, and re-deployed to the links below, but I seem to get the same bad behavior (but maybe inconsistently? -- 00-setup seems to work).

Can you guys confirm?

https://tutorials.shinyapps.io/00-setup/
https://tutorials.shinyapps.io/01-data-basics/
https://tutorials.shinyapps.io/03a-data-manip-filter/
https://tutorials.shinyapps.io/03b-data-manip-mutate/
https://tutorials.shinyapps.io/03c-data-manip-summarise/

**Comment by robbyshaver** [comment]
Just took a look - the code running is more up to date and I am seeing tutor firing a section-skipped event on an undefined section when I first open some of the tutorials (3a and 3b).

However, the code does not include the possible fix I checked in a an hour or so ago, so if you could get the latest and repost, that would be helpful.

I'm also seeing occasional internal server errors when I open 01.

**Comment by jjallaire** [comment]
I saw those internal server errors as well, I'm not sure what that is. I'll
post a bug for it and investigate later.

**Comment by garrettgman** [comment]
I re-installed tutor at 7:22 tonight after your messages and re-rendered, and re-deployed each tutorial. It seems to me that the Continue buttons may now work correctly.

**Comment by jjallaire** [comment]
I'll close this as it looks like the issues have been resolved. @garrettgman Please re-open if you see additional problems.

**Comment by garrettgman** [comment]
I've encountered this behavior again: the continue buttons  do not work when I deploy to shinyapps.io_ (continue buttons always work locally).

I may have identified a cause:

If I deploy a tutorial after clicking "Start Over", or before completing any sections, the tutorial works correctly on shinyapps.io. For example, I deployed this tutorial in this way and it works correctly: https://tutorials.shinyapps.io/04-programming-basics. Here's how it looked before I deployed it

<img width="1199" alt="screen shot 2017-06-12 at 11 29 43 am" src="https://user-images.githubusercontent.com/85014/27041656-9589f2a8-4f62-11e7-8e7d-b621afe7054a.png">

If I deploy the tutorial after completing sections and without clicking "Start Over", (e.g. with a history), the tutorial does not work on shinyapps.io. For example, I redeployed the same tutorial after completing a few sections here: https://tutorials.shinyapps.io/04-programming-basics-alt. In this version, the continue buttons do not work, with the exception of the continue button at the end of the first section (Welcome), which is actually a disguised "Next Section" button. Here's how the tutorial looked before I deployed it

<img width="1199" alt="screen shot 2017-06-12 at 11 29 37 am" src="https://user-images.githubusercontent.com/85014/27041690-a8592b38-4f62-11e7-9364-11d78120b0a8.png">

Note: I can sometimes get the continue buttons to work in the broken tutorial by jumping around between sections within the tutorial.

**TL;DR** - It seems that you need to erase a local tutorials history by clicking "Start over" before you deploy it to shinyapps.io. If you do not, the continue buttons will not work when deployed.

**Comment by garrettgman** [comment]
An addendum to the last comment: the continue buttons for the good link above have stopped working for me on subsequent visits (https://tutorials.shinyapps.io/04-programming-basics).

I also notice that the tutorial does not seem to remember my progress when I revisit it later.

**Comment by jjallaire** [comment]
It's possible that this is a shinyapps.io error. @jspiewak I see the following 500 error when revisiting the tutorial on shinyappsio (this doesn't occur locally):
<img width="1409" alt="screen shot 2017-06-13 at 7 48 31 am" src="https://user-images.githubusercontent.com/104391/27081090-e2c5af9e-500c-11e7-93c3-855a3d03bbff.png">

@garrettgman This 500 error would explain all of the other issues you are seeing regarding revisiting of tutorials.

**Comment by jjallaire** [comment]
I've confirmed that this also occurs on beta.rstudioconnect.com:
<img width="1410" alt="screen shot 2017-06-13 at 7 59 47 am" src="https://user-images.githubusercontent.com/104391/27081440-6529a66a-500e-11e7-9bb1-c2da4e56d267.png">

Simple repro Rmd is:

    ---
    title: "Tutorial"
    output: learnr::tutorial
    runtime: shiny_prerendered
    ---
    
    ```{r setup, include=FALSE}
    library(learnr)
    knitr::opts_chunk$set(echo = FALSE)
    ```

    ## Topic 1

    ### Exercise 

    *Here's a simple exercise with an empty code chunk provided for entering the answer.*

    Write the R code required to add two plus two:

    ```{r two-plus-two, exercise=TRUE}
    
    ```

This is AFAIK a new error condition. Perhaps a change in a recent version of RSC or (less likely so) Shiny? @trestletech Could you take a look at this as well?

**Comment by robbyshaver** [comment]
I don't know if this is related or not, but when I try to view the tutorial at: https://tutorials.shinyapps.io/04-programming-basics, I get the following error:

![image](https://user-images.githubusercontent.com/1553877/27085379-133297b8-501d-11e7-9e40-c41741554248.png)

**Comment by jspiewak** [comment]
@robbyshaver unrelated, that is bug with lucid where it doesn't deal with incorrectly cased applications. The actual URL is https://tutorials.shinyapps.io/04-Programming-Basics/

**Comment by garrettgman** [comment]
@robbyshaver I've seen that before, but thought it was my web browser autofill misfiring. When I retyped the URL it worked.

**Comment by garrettgman** [comment]
@jspiewak Should all application names be lower case? Good to know

**Comment by jspiewak** [comment]
Well, we should fix the bug for starters.
The work around is to use the same case in the URL as when the application was created.
If one always lower cases application names, then one doesn't need to think about it.
But really we should just fix the bug already. :)

**Comment by jjallaire** [comment]
I now have a local repro for the 500 error (it has to do with how jsonlite is round-tripping R data structures, perhaps something changed in the v1.5 version of jsonlite). More soon!

**Comment by jjallaire** [comment]
This should now be fixed. @garrettgman Could you confirm that all is well on your end?

**Comment by garrettgman** [comment]
Thanks guys. I can no longer reproduce the error.

@jjallaire I do notice two things. Let me know if you'd like me to open them as new issues:

1. There is a noticeable delay between when a tutorial loads and when it advances to the saved state. 

2. I deployed the same tutorial in two places (to make the comparison above). They seem to share state now (each advances to where I left off in the other).
https://tutorials.shinyapps.io/04-programming-basics/
https://tutorials.shinyapps.io/04-programming-basics-alt/

**Comment by jjallaire** [comment]
What browser are you using when you see the shared state behavior?

Could you send me the source code for that tutorial to aid in reproducding?

On Thu, Jun 15, 2017 at 3:11 PM, Garrett Grolemund <notifications@github.com

**Comment by garrettgman** [comment]
@jjallaire, I am using Google Chrome. Here is the source code:
https://github.com/rstudio/ggtutorials/blob/master/the-basics/04-Programming-Basics/04-Programming-Basics.Rmd

**Comment by jjallaire** [comment]
Okay, found the problem. All should be well! (please test your tutorials and let me know if you see any problems)

**Comment by garrettgman** [comment]
Ferit observed the same behavior today (10/17/2018) while working through the https://rstudio.cloud/learn/primers/2.3 tutorial, i.e.

The second Continue button in the tutorial did not advance the tutorial in Google Chrome (the first Continue button is a Next Topic button). The Continue button began to work when he revisited the tutorial later.

**Comment by schloerke** [comment]
This seems to be working. Closing

**Comment by amf60** [comment]
R studio Cloud - Primer - Iterate - "Continue" button issue - 

despite having correctly answered the exercise the progress to next part is blocked by a pop window saying "You must complete the exercise in this section before continuing". Forced the way to the next part and got the same issue with the other sessions of iterate. So I am uncertain of the real progress done with this very important lesson. Thank you in advance for your help.

![Screenshot 2020-09-16 at 14 42 20](https://user-images.githubusercontent.com/39899585/93346001-65988d00-f82b-11ea-8083-915ad234052b.png)
